pageserver/src/tenant_mgr.rs

```diff
@@ -348,25 +386,90 @@ pub fn list_tenants() -> Vec<(TenantId, TenantState)> {
+where
+    F: FnOnce(hash_map::VacantEntry<TenantId, Arc<Tenant>>) -> anyhow::Result<V>,
+{
     match tenants_state::write_tenants().entry(tenant_id) {
         hash_map::Entry::Occupied(e) => {
-            // Cannot attach a tenant that already exists. The error message depends on
-            // the state it's in.
-            match e.get().current_state() {
-                TenantState::Attaching => {
-                    anyhow::bail!("tenant {tenant_id} attach is already in progress")
-                }
-                current_state => {
-                    anyhow::bail!("tenant already exists, current state: {current_state:?}")
+            anyhow::bail!(
+                "tenant {tenant_id} already exists, state: {:?}",
+                e.get().current_state()
+            )
+        }
+        hash_map::Entry::Vacant(v) => run(v),
+    }
+}
+
+/// Stops and removes the tenant from memory, if it's not [`TenantState::Stopping`] already, bails otherwise.
+/// Allows to remove other tenant resources manually, via `tenant_cleanup`.
+/// If the cleanup fails, tenant will stay in memory in [`TenantState::Broken`] state, and another removal
+/// operation would be needed to remove it.
+async fn remove_tenant_from_memory<V, F>(
+    tenant_id: TenantId,
+    tenant_cleanup: F,
+) -> anyhow::Result<V>
+where
+    F: std::future::Future<Output = anyhow::Result<V>>,
+{
+    // XXX:
+    // It's important to keep the tenant in memory after the final cleanup, to avoid cleanup races.
+    // The exclusive lock here ensures we don't miss the stats updates before trying another removal.
+    // tenant-wde cleanup operations may take some time (removing the entire tenant directory), we want to
+    // avoid holding the lock for the entire process.
+    {
+        let tenants_accessor = tenants_state::write_tenants();
+        match tenants_accessor.get(&tenant_id) {
+            Some(tenant) => match tenant.current_state() {
+                TenantState::Attaching
+                | TenantState::Loading
+                | TenantState::Broken
+                | TenantState::Active => tenant.set_stopping(),
+                TenantState::Stopping => {
+                    anyhow::bail!("Tenant {tenant_id} is stopping already")
                 }
+            },
+            None => anyhow::bail!("Tenant not found for id {tenant_id}"),
+        }
+    }
+
+    // shutdown all tenant and timeline tasks: gc, compaction, page service)
+    task_mgr::shutdown_tasks(None, Some(tenant_id), None).await;
+
+    match tenant_cleanup
+        .await
+        .with_context(|| format!("Failed to run cleanup for tenant {tenant_id}"))
+    {
+        Ok(hook_value) => {
+            let mut tenants_accessor = tenants_state::write_tenants();
+            if tenants_accessor.remove(&tenant_id).is_none() {
+                warn!("Tenant {tenant_id} got removed from memory before operation finished");
```

Comment:
So, can this even happen?
The point of the `match` at the beginning of the function is that we only admit exactly on tokio task into the remainder of this function. All others are locked out because the tenant state is `Stopping` then.

That's necessary to serialize, say, concurrent detach & ignore.

If we arrive here, we failed to do that.


The only other function that removes entries from the tenants map is `shutdown_all_tenants`.
I guess we could still race with that one.
Any harm in making it use this function here as well, then turn this `warn` into an `assert`?

Reply:
I'm not sure it's worth the code changes: technically, now there's a very small race window for `shutdown_all_tenants` with anything, since before it's called, all pageserver endpoints are stopped.

Also, this function is somewhat special since tries to operate on tenants in bulk, infallible.
In particular, it cleans the tenant map, which is not working well with our model of synchronization, but does work due to the fact that we've stopped all other tasks.

I think, warning is ok here since will have a small chance to appear only when shutdown was called during some slow `tenant_cleanup` call which should be rare and visible in logs thanks to such messages.

Reply:
Ok, let's ship this.